reference.md

```diff
@@ -0,0 +1,381 @@
+
+* [Overview](#weakref-overview)
+* [WeakRef API](#weakref-api)
+    * [The `WeakRef` Constructor](#the-weakref-constructor)
+    * [`WeakRef.prototype.deref`](#weakrefprototypederef)
+* [Notes](#notes-on-weakrefs)
+
+## WeakRef Overview
+
+To create a WeakRef, you use the `WeakRef` constructor, passing in the target object (the object you want a weak reference to):
+
+```js
+let ref = new WeakRef(someObject);
+```
+
+Then at some point you stop using `someObject` (it goes out of scope, etc.) while keeping the WeakRef instance (`ref`). At that point, to get the object from a WeakRef instance, use its `deref` method:
+
+```js
+let obj = ref.deref();
+if (obj) {
+    // ...use `obj`...
+}
+```
+
+`deref` returns the object, or `undefined` if the object is no longer available.
+
+## WeakRef API
+
+### The `WeakRef` Constructor
+
+The `WeakRef` constructor creates a WeakRef instance for the given `target` object:
+
+```js
+ref = new WeakRef(target)
+```
+
+**Parameters**
+
+* `target` - The target object for the WeakRef instance.
+
+**Returns**
+
+* The WeakRef instance.
+
+**Notes**
+
+* `WeakRef` is a constructor function, so it throws if called as a function rather than as a constructor.
+* `WeakRef` is designed to be subclassed, and so can appear in the `extends` of a `class` definition and similar.
+
+### `WeakRef.prototype.deref`
+
+The `deref` method returns the WeakRef instance's target object, or `undefined` if the target object has been reclaimed:
+
+```js
+targetOrUndefined = weakRef.deref();
+```
+
+**Parameters**
+
+* *(none)*
+
+**Returns**
+
+* The target object of the WeakRef, or `undefined`.
+
+**Errors**
+
+* *(none)*
+
+## Notes on WeakRefs
+
+Some notes on WeakRefs:
+
+* If your code has just created a WeakRef for a target object, or has gotten a target object from a WeakRef's `deref` method, that target object will not be reclaimed until the end of the current JavaScript [job](https://tc39.es/ecma262/#job) (including any promise reaction jobs that run at the end of a script job). That is, you can only "see" an object get reclaimed between turns of the event loop. This is primarily to avoid making the behavior of any given JavaScript engine's garbage collector apparent in code&nbsp;&mdash; because if it were, people would write code relying on that behavior, which would break when the garbage collector's behavior changed. (Garbage collection is a hard problem; JavaScript engine implementers are constantly refining and improving how it works.)
+* If multiple WeakRefs have the same target, they're consistent with one another. The result of calling `deref` on one of them will match the result of calling `deref` on another of them (in the same job), you won't get the target object from one of them but `undefined` from another.
+* If the target of a WeakRef is also in a [`FinalizationRegistry`](#finalizationregistry), the WeakRef's target is cleared *before* any cleanup callback associated with the registry is called.
```

Comment:
before or at the same time. Maybe it'd be more clear to put this backwards, and as part of the FinalizationGroup notes: it won't be possible, during a finalizer callback, to read a WeakRef and get the object that was registered--these Weakefs will always be undefined.

Reply:
I'm happy to move it. :-) But it literally is *before*, right? In [Execution](https://tc39.es/proposal-weakrefs/#sec-weakref-execution), for a given `obj`, the agent clears any [[WeakRefTarget]]s of WeakRefs before moving on to finalization for it (Step 1.a comes before Step 1.b). I find it easier to explain and reason about this if I know that when `obj` is going away / gone, any WeakRefs for it are cleared before any cleanup callbacks for it occur. The "before" here doesn't refer to event turns or anything, just the boring old temporal meaning. :-)

Reply:
(I have the other fixes ready to go -- can't believe I missed a couple of those -- just waiting to resolve this point.)

Reply:
Those steps all happen "atomically", so it's not really before.